quinn-proto/src/endpoint.rs

```diff
@@ -1037,36 +1131,46 @@ impl Incoming {
     ///
     /// This has the same behavior as [`Connection::local_ip`]
     pub fn local_ip(&self) -> Option<IpAddr> {
-        self.addresses.local_ip
+        self.0.as_ref().unwrap().addresses.local_ip
     }
 
     /// The peer's UDP address.
     pub fn remote_address(&self) -> SocketAddr {
-        self.addresses.remote
+        self.0.as_ref().unwrap().addresses.remote
     }
 
     /// Whether the socket address that is initiating this connection has been validated.
     ///
     /// This means that the sender of the initial packet has proved that they can receive traffic
     /// sent to `self.remote_address()`.
     pub fn remote_address_validated(&self) -> bool {
-        self.retry_src_cid.is_some()
+        self.0.as_ref().unwrap().retry_src_cid.is_some()
     }
 }
 
 impl fmt::Debug for Incoming {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let inner = self.0.as_ref().unwrap();
         f.debug_struct("Incoming")
-            .field("addresses", &self.addresses)
-            .field("ecn", &self.ecn)
+            .field("addresses", &inner.addresses)
+            .field("ecn", &inner.ecn)
             // packet doesn't implement debug
             // rest is too big and not meaningful enough
-            .field("retry_src_cid", &self.retry_src_cid)
-            .field("orig_dst_cid", &self.orig_dst_cid)
+            .field("retry_src_cid", &inner.retry_src_cid)
+            .field("orig_dst_cid", &inner.orig_dst_cid)
+            .field("incoming_idx", &inner.incoming_idx)
             .finish_non_exhaustive()
     }
 }
 
+impl Drop for Incoming {
+    fn drop(&mut self) {
+        if self.0.is_some() {
```

Comment:
Is this the only reason we've wrapped the contents in an `Option`? I'd rather avoid the readability cost of traversing and `unwrap`ing the `Option` everywhere. Maybe add a ZST field that logs on drop, and `mem::forget` it on the intended path?

Reply:
Good idea.

Reply:
Done.